### PR TITLE
JMS maxAckInterval

### PR DESCRIPTION
**Pull request by manuelcueto**
References #2583 .
Adds timeout for flushing acks back to the JMS broker instead on relying only on buffer completion.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerSettings.scala [hunk 1]
```diff
@@ -206,7 +222,7 @@ object JmsConsumerSettings {
   /**
    * Java API: Reads from the default config provided by the actor system at `alpakka.jms.consumer`.
    *
-   * @param actorSystem The actor system
+   * @param actorSystem       The actor system
```
Comment (seglo) [inline jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerSettings.scala:225]: I think your IDE is doing some reformatting.

File: jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala [hunk 2]
```diff
@@ -214,6 +215,12 @@ private[jms] trait JmsConnector[S <: JmsSession] {
   }
 
   override def onTimer(timerKey: Any): Unit = timerKey match {
+    case FlushAcknowledgementsTimerKey(session, timeout) =>
```
Comment (seglo) [inline jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala:218]: This timer should be cancelled when the stage shuts down (or a session shuts down?)

File: jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala [hunk 3]
```diff
@@ -394,7 +401,7 @@ object JmsConnector {
   case object TimedOut extends ConnectionAttemptStatus
 
   case class AttemptConnect(attempt: Int, backoffMaxed: Boolean)
-
+  case class FlushAcknowledgementsTimerKey(jmsSession: JmsAckSession, timeout: FiniteDuration)
```
Comment (seglo) [inline jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala:404]:
```suggestion
  final case class FlushAcknowledgementsTimerKey(jmsSession: JmsAckSession, timeout: FiniteDuration)
```

File: jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala [hunk 4]
```diff
@@ -394,7 +401,7 @@ object JmsConnector {
   case object TimedOut extends ConnectionAttemptStatus
 
   case class AttemptConnect(attempt: Int, backoffMaxed: Boolean)
```
Comment (seglo) [inline jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala:403]:
```suggestion
  final case class AttemptConnect(attempt: Int, backoffMaxed: Boolean)
```

**Review by seglo — COMMENTED** [review]
Thanks for working on this. A few comments.

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerSettings.scala (hunk 1 above)
Comment (manuelcueto) [reply jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerSettings.scala:225]: Yes, I've run compile and scalafmtAll from  sbt and it seems to be ok with this formatting. Would you rather i just restore the file as it was before?

**Comment by seglo** [comment]
Looks good @manuelcueto . I think all that's left is to update the JMS documentation to mention this configuration.

File: jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerSettings.scala (hunk 1 above)
Comment (seglo) [reply jms/src/main/scala/akka/stream/alpakka/jms/JmsConsumerSettings.scala]: If the code style sbt task is happy then I guess it's fine. I just wanted to point it out in case you had IDE formatting things enabled.

File: jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala [hunk 5]
```diff
@@ -364,7 +366,7 @@ class JmsAckConnectorsSpec extends JmsSpec {
     }
 
     "shutdown when waiting to acknowledge messages" in withServer() { server =>
-      val connectionFactory = server.createConnectionFactory
+      val connectionFactory = server.createQueueConnectionFactory
```
Comment (seglo) [inline jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala:369]: What's the purpose of this change?

File: jms/src/main/resources/reference.conf [hunk 6]
```diff
@@ -53,6 +52,10 @@ alpakka.jms {
     # For use with transactions, if true the stream fails if Alpakka rolls back the transaction
     # when `ack-timeout` is hit.
     fail-stream-on-ack-timeout = false
+    # Max interval before sending queued acknowledges back to the broker. (Used by AckSources.)
+    # max-ack-interval = 5 seconds
+    # Max number of acks queued by AckSource before they are sent to broker. (Unless MaxAckInterval is specified).
+    max-pending-acks = 100
```
Comment (seglo) [inline jms/src/main/resources/reference.conf:58]:
```suggestion
    max-pending-acks = ${alpakka.jms.consumer.buffer-size}
```

This will let the change be transparent to users who used `buffer-size` to control the pending acks in previous versions.

File: docs/src/main/paradox/jms/consumer.md [hunk 7]
```diff
@@ -94,6 +96,7 @@ The `sessionCount` parameter controls the number of JMS sessions to run in paral
 *  Using multiple sessions increases throughput, especially if acknowledging message by message is desired.
 *  Messages may arrive out of order if `sessionCount` is larger than 1.
 *  Message-by-message acknowledgement can be achieved by setting `bufferSize` to 0, thus disabling buffering. The outstanding messages before backpressure will be the `sessionCount`.
+*  If buffering is enabled, setting `maxAckInterval` will mean acknowledgment happens at most `maxAckInterval` time after `ackEnvelope.acknowledge()` is called. If not set, acknowledgment will not reach the broker until a new message arrives.
```
Comment (seglo) [inline docs/src/main/paradox/jms/consumer.md:99]: Maybe rephrase this a bit

ex)

> If buffering is enabled then it's possible for messages to remain in the buffer and never be acknowledged (or acknowledged after a long time) when no new elements arrive to reach the `maxPendingAcks` threshold. By setting `maxAckInterval` messages will be acknowledged after the defined interval _or_ number of pending acks, whichever comes first.

File: jms/src/test/scala/docs/scaladsl/JmsBufferedAckConnectorsSpec.scala [hunk 8]
```diff
@@ -380,5 +382,47 @@ class JmsBufferedAckConnectorsSpec extends JmsSharedServerSpec {
       }
       resultList.toSet should contain theSameElementsAs numsIn.map(_.toString)
     }
+
+    "flush acknowledgments to broker after flush.timeout triggers" in withConnectionFactory() { connectionFactory =>
```
Comment (seglo) [inline jms/src/test/scala/docs/scaladsl/JmsBufferedAckConnectorsSpec.scala:386]: Give this test a more meaningful name now that we've settled on terms for max-ack-interval.

**Review by seglo — COMMENTED** [review]
Looking good. A few more things :)

File: jms/src/main/resources/reference.conf (hunk 6 above)
Comment (manuelcueto) [reply jms/src/main/resources/reference.conf:58]: Nice

File: jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala (hunk 5 above)
Comment (manuelcueto) [reply jms/src/test/scala/akka/stream/alpakka/jms/scaladsl/JmsAckConnectorsSpec.scala:369]: I honestly think i was trying to change that in the test i introduced, and just missed 😓

File: docs/src/main/paradox/jms/consumer.md (hunk 7 above)
Comment (manuelcueto) [reply docs/src/main/paradox/jms/consumer.md:99]: I like it better yeah.

**Review by seglo — APPROVED** [review]
LGTM. Nice cleanup :+1:
